src/api/cmd/chart-repo-sync/sync_test.go

```diff
@@ -0,0 +1,447 @@
+    version: 0.7.5
+  - appVersion: 4.9.0
+    created: 2017-12-01T11:49:00.136950565Z
+    description: Web publishing platform for building blogs and websites.
+    digest: a69139ef3008eeb11ca60261ec2ded61e84ce7db32bb3626056e84bcff7ec270
+    engine: gotpl
+    home: http://www.wordpress.com/
+    icon: https://bitnami.com/assets/stacks/wordpress/img/wordpress-stack-220x234.png
+    keywords:
+    - wordpress
+    - cms
+    - cms
+    - blog
+    - http
+    - web
+    - application
+    - php
+    maintainers:
+    - email: containers@bitnami.com
+      name: bitnami-bot
+    name: wordpress
+    sources:
+    - https://github.com/bitnami/bitnami-docker-wordpress
+    urls:
+    - https://kubernetes-charts.storage.googleapis.com/wordpress-0.7.4.tgz
+    version: 0.7.4
+`
+
+var invalidRepoIndexYAML = "invalid"
+
+type badHTTPClient struct{}
+
+func (h *badHTTPClient) Do(req *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+	w.WriteHeader(500)
+	return w.Result(), nil
+}
+
+type goodHTTPClient struct{}
+
+func (h *goodHTTPClient) Do(req *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+	// Don't accept trailing slashes
+	if strings.HasPrefix(req.URL.Path, "//") {
+		w.WriteHeader(500)
+	}
+	// Ensure we're sending the right User-Agent
+	if !strings.Contains(req.Header.Get("User-Agent"), "chart-repo-sync") {
+		w.WriteHeader(500)
+	}
+	w.Write([]byte(validRepoIndexYAML))
+	return w.Result(), nil
+}
+
+type badIconClient struct{}
+
+func (h *badIconClient) Do(req *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+	w.Write([]byte("not-an-image"))
+	return w.Result(), nil
+}
+
+type goodIconClient struct{}
+
+func iconBytes() []byte {
+	var b bytes.Buffer
+	img := imaging.New(1, 1, color.White)
+	imaging.Encode(&b, img, imaging.PNG)
+	return b.Bytes()
+}
+
+func (h *goodIconClient) Do(req *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+	w.Write(iconBytes())
+	return w.Result(), nil
+}
+
+type goodTarballClient struct {
+	c          chart
+	skipReadme bool
+}
+
+var testChartReadme = "# readme for chart\n\nBest chart in town"
+
+func (h *goodTarballClient) Do(req *http.Request) (*http.Response, error) {
+	w := httptest.NewRecorder()
+	gzw := gzip.NewWriter(w)
+	files := []tarballFile{{h.c.Name + "/Chart.yaml", "should be a Chart.yaml here..."}}
+	if !h.skipReadme {
+		files = append(files, tarballFile{h.c.Name + "/README.md", testChartReadme})
+	}
+	createTestTarball(gzw, files)
+	gzw.Flush()
+	return w.Result(), nil
+}
+
+func Test_syncURLInvalidity(t *testing.T) {
+	tests := []struct {
+		name    string
+		repoURL string
+	}{
+		{"invalid URL", "not-a-url"},
+		{"invalid URL", "https//google.com"},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			err := sync("test", tt.repoURL)
+			assert.ExistsErr(t, err, tt.name)
+		})
+	}
+}
+
+func Test_fetchRepoIndex(t *testing.T) {
```

Comment:
I am curious, why are you using the underscore after the Test required word? Is this a new GO idiom?

Reply:
I believe so for internal/unexposed function names. VSCode automatically creates these names.